src/twisted/application/test/test_internet.py

```diff
@@ -1031,3 +1031,78 @@ def test_stopServiceOnStoppedService(self):
 
         self.assertIsNone(self.successResultOf(firstStopDeferred))
         self.assertIsNone(self.successResultOf(secondStopDeferred))
+
+
+    def test_onNewConnectionCalledWhenServiceStarts(self):
+        """
+        The C{onNewConnection} callable is called when the
+        L{ClientService.startService} is called.
```

Comment:
Is this true?
I was expecting onNewConnection to be called only when a connection is done.

If you call ClientService.startService and the connection fails, onNewConnection is not called.

So maybe

onNewConnection is called after ClientService.startService once the connection is made.